src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/RawConnectionStream.cs

```diff
@@ -65,14 +117,63 @@ public override Task WriteAsync(byte[] buffer, int offset, int count, Cancellati
                 return WriteAsync(new ReadOnlyMemory<byte>(buffer, offset, count), cancellationToken);
             }
 
-            public override Task WriteAsync(ReadOnlyMemory<byte> source, CancellationToken cancellationToken = default) =>
-                _connection == null ? Task.FromException(new IOException(SR.net_http_io_write)) :
-                source.Length > 0 ? _connection.WriteWithoutBufferingAsync(source, cancellationToken) :
-                Task.CompletedTask;
+            public override Task WriteAsync(ReadOnlyMemory<byte> source, CancellationToken cancellationToken)
+            {
+                if (cancellationToken.IsCancellationRequested)
+                {
+                    return Task.FromCanceled(cancellationToken);
+                }
+
+                if (_connection == null)
+                {
+                    return Task.FromException(new IOException(SR.net_http_io_write));
+                }
+
+                if (source.Length == 0)
+                {
+                    return Task.CompletedTask;
+                }
+
+                Task writeTask = _connection.WriteWithoutBufferingAsync(source);
+                return writeTask.IsCompleted ?
+                    writeTask :
+                    WaitWithConnectionCancellationAsync(writeTask, cancellationToken);
+            }
+
+            public override Task FlushAsync(CancellationToken cancellationToken)
+            {
+                if (cancellationToken.IsCancellationRequested)
+                {
+                    return Task.FromCanceled(cancellationToken);
+                }
+
+                if (_connection == null)
+                {
+                    return Task.CompletedTask;
+                }
+
+                Task flushTask = _connection.FlushAsync();
+                return flushTask.IsCompleted ?
+                    flushTask :
+                    WaitWithConnectionCancellationAsync(flushTask, cancellationToken);
+            }
 
-            public override Task FlushAsync(CancellationToken cancellationToken) =>
-                _connection != null ? _connection.FlushAsync(cancellationToken) :
-                Task.CompletedTask;
+            private async Task WaitWithConnectionCancellationAsync(Task task, CancellationToken cancellationToken)
```

Comment:
It seems like this could be used in other places as well, no?  Is the issue just avoiding the extra overhead for another async function?

Reply:
> Is the issue just avoiding the extra overhead for another async function?

Right.  If we had something like https://github.com/dotnet/roslyn/issues/15491, then we would definitely reuse it in more places.